x/cronos/keeper/evm.go

```diff
@@ -0,0 +1,154 @@
+package keeper
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	evmtypes "github.com/tharsis/ethermint/x/evm/types"
+
+	"github.com/crypto-org-chain/cronos/x/cronos/types"
+)
+
+// CallEVM execute an evm message from native module
+func (k Keeper) CallEVM(ctx sdk.Context, to *common.Address, data []byte, value *big.Int) (*ethtypes.Message, *evmtypes.MsgEthereumTxResponse, error) {
+	k.evmKeeper.WithContext(ctx)
+
+	nonce := k.evmKeeper.GetNonce(types.EVMModuleAddress)
+	msg := ethtypes.NewMessage(
+		types.EVMModuleAddress,
+		to,
+		nonce,
+		value,         // amount
+		25000000,      // gasLimit, FIXME configuration
+		big.NewInt(0), // gasPrice
+		data,
+		nil,   // accessList
+		false, // checkNonce
+	)
+
+	params := k.evmKeeper.GetParams(ctx)
+	// return error if contract creation or call are disabled through governance
+	if !params.EnableCreate && to == nil {
+		return nil, nil, errors.New("failed to create new contract")
+	} else if !params.EnableCall && to != nil {
+		return nil, nil, errors.New("failed to call contract")
+	}
+	ethCfg := params.ChainConfig.EthereumConfig(k.evmKeeper.ChainID())
+
+	// get the coinbase address from the block proposer
+	coinbase, err := k.evmKeeper.GetCoinbaseAddress(ctx)
+	if err != nil {
+		return nil, nil, errors.New("failed to obtain coinbase address")
+	}
+	evm := k.evmKeeper.NewEVM(msg, ethCfg, params, coinbase, nil)
+	ret, err := k.evmKeeper.ApplyMessage(evm, msg, ethCfg, true)
+	if err != nil {
+		return nil, nil, err
+	}
+	k.evmKeeper.CommitCachedContexts()
+	return &msg, ret, nil
+}
+
+// CallCronosERC20 call a method of CronosERC20 contract
+func (k Keeper) CallCronosERC20(ctx sdk.Context, contract common.Address, method string, args ...interface{}) ([]byte, error) {
+	data, err := types.CronosERC20Contract.ABI.Pack(method, args...)
+	if err != nil {
+		return nil, err
+	}
+	_, res, err := k.CallEVM(ctx, &contract, data, big.NewInt(0))
+	if err != nil {
+		return nil, err
+	}
+	if res.Failed() {
+		return nil, fmt.Errorf("call contract failed: %s, %s, %s", contract.Hex(), method, res.Ret)
+	}
+	return res.Ret, nil
+}
+
+// DeployCronosERC20 deploy an embed erc20 contract
+func (k Keeper) DeployCronosERC20(ctx sdk.Context, denom string) (common.Address, error) {
```

Comment:
`DeployModuleContract` to make it more generic? (ERC20 might be changed to CRO20)

Reply:
This function is to deploy a specific contract named CronosERC20, so to change the name, we better change that contract name systematically. `CronosToken`?

Reply:
How about `DeployModuleErc20` or `DeployModuleToken` ?

What bother me is the term `Cronos` because it could refers to the blockchain or the app, not necessarly the the module which is one owning the contract

Reply:
Hmm, `ModuleErc20` sounds good, we can change the erc20 to cro20(or whatever) later after we made the decision.

Reply:
By the way, ERC20 -> CRC20 renaming is done too.